From 3f249104f5eb9654786cb768489ed6c230527567 Mon Sep 17 00:00:00 2001
From: Jason Dere <jdere@hortonworks.com>
Date: Thu, 1 Dec 2016 15:25:54 -0800
Subject: [PATCH 0851/1164] CDH-47620: HIVE-15291: Comparison of timestamp
 fails if only date part is provided (Dhiraj
 Kumar, reviewed by Jason Dere)

(cherry picked from commit d53aa79caf9f1574a05a98d8b899ca13a5c5b24f)

Change-Id: Ic9b8123008e0d1a15c9a0aad9dbd3d986691495b
---
 .../queries/clientpositive/timestamp_date_only.q   |    3 +++
 .../clientpositive/timestamp_date_only.q.out       |   27 ++++++++++++++++++++
 .../primitive/PrimitiveObjectInspectorUtils.java   |    3 +++
 .../TestPrimitiveObjectInspectorUtils.java         |    7 +++++
 4 files changed, 40 insertions(+)
 create mode 100644 ql/src/test/queries/clientpositive/timestamp_date_only.q
 create mode 100644 ql/src/test/results/clientpositive/timestamp_date_only.q.out

diff --git a/ql/src/test/queries/clientpositive/timestamp_date_only.q b/ql/src/test/queries/clientpositive/timestamp_date_only.q
new file mode 100644
index 0000000..324938e
--- /dev/null
+++ b/ql/src/test/queries/clientpositive/timestamp_date_only.q
@@ -0,0 +1,3 @@
+select cast("2016-12-29 23:59:59" as timestamp) < "2016-12-30";
+select cast("2016-12-30 00:00:00" as timestamp) = "2016-12-30";
+select cast("2016-12-30 00:00:01" as timestamp) > "2016-12-30";
\ No newline at end of file
diff --git a/ql/src/test/results/clientpositive/timestamp_date_only.q.out b/ql/src/test/results/clientpositive/timestamp_date_only.q.out
new file mode 100644
index 0000000..8f37a29
--- /dev/null
+++ b/ql/src/test/results/clientpositive/timestamp_date_only.q.out
@@ -0,0 +1,27 @@
+PREHOOK: query: select cast("2016-12-29 23:59:59" as timestamp) < "2016-12-30"
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+POSTHOOK: query: select cast("2016-12-29 23:59:59" as timestamp) < "2016-12-30"
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+true
+PREHOOK: query: select cast("2016-12-30 00:00:00" as timestamp) = "2016-12-30"
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+POSTHOOK: query: select cast("2016-12-30 00:00:00" as timestamp) = "2016-12-30"
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+true
+PREHOOK: query: select cast("2016-12-30 00:00:01" as timestamp) > "2016-12-30"
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+POSTHOOK: query: select cast("2016-12-30 00:00:01" as timestamp) > "2016-12-30"
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+true
diff --git a/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java b/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java
index 5ccacf1..be2d950 100644
--- a/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java
+++ b/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java
@@ -1095,6 +1095,9 @@ static Timestamp getTimestampFromString(String s) {
         s = s.substring(0, periodIdx + 10);
       }
     }
+    if (s.indexOf(' ') < 0) {
+      s = s.concat(" 00:00:00");
+    }
     try {
       result = Timestamp.valueOf(s);
     } catch (IllegalArgumentException e) {
diff --git a/serde/src/test/org/apache/hadoop/hive/serde2/objectinspector/primitive/TestPrimitiveObjectInspectorUtils.java b/serde/src/test/org/apache/hadoop/hive/serde2/objectinspector/primitive/TestPrimitiveObjectInspectorUtils.java
index 021e670..6fd50bd 100644
--- a/serde/src/test/org/apache/hadoop/hive/serde2/objectinspector/primitive/TestPrimitiveObjectInspectorUtils.java
+++ b/serde/src/test/org/apache/hadoop/hive/serde2/objectinspector/primitive/TestPrimitiveObjectInspectorUtils.java
@@ -18,6 +18,8 @@
 
 package org.apache.hadoop.hive.serde2.objectinspector.primitive;
 
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
 import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector.PrimitiveCategory;
 import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorUtils.PrimitiveGrouping;
 
@@ -60,4 +62,9 @@ public void testGetPrimitiveGrouping() {
     assertEquals(PrimitiveGrouping.VOID_GROUP,
         PrimitiveObjectInspectorUtils.getPrimitiveGrouping(PrimitiveCategory.VOID));
   }
+
+  public void testGetTimestampFromString() {
+    DateFormat localDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
+    assertEquals("2015-02-07 00:00:00.000", localDateFormat.format(PrimitiveObjectInspectorUtils.getTimestampFromString("2015-02-07")));
+  }
 }
-- 
1.7.9.5

